narwhals/_pandas_like/group_by.py

```diff
@@ -1,24 +1,218 @@
 from __future__ import annotations
 
-import collections
 import warnings
-from typing import TYPE_CHECKING, Any, ClassVar
+from functools import lru_cache
+from itertools import chain
+from operator import methodcaller
+from typing import TYPE_CHECKING, Any, ClassVar, Literal, overload
 
 from narwhals._compliant import EagerGroupBy
 from narwhals._expression_parsing import evaluate_output_names_and_aliases
-from narwhals._pandas_like.utils import select_columns_by_name
-from narwhals._utils import find_stacklevel
+from narwhals._pandas_like.utils import (
+    is_nullable_dtype_backend,
+    native_to_narwhals_dtype,
+)
+from narwhals._typing_compat import TypeVar
+from narwhals._utils import find_stacklevel, generate_temporary_column_name
 
 if TYPE_CHECKING:
-    from collections.abc import Iterator, Mapping, Sequence
+    from collections.abc import Callable, Iterable, Iterator, Mapping, Sequence
 
-    from narwhals._compliant.typing import NarwhalsAggregation
+    import pandas as pd
+    from pandas.api.typing import (
+        DataFrameGroupBy as _NativeGroupBy,
+        SeriesGroupBy as _SeriesGroupBy,
+    )
+    from typing_extensions import TypeAlias, Unpack
+
+    from narwhals._compliant.typing import NarwhalsAggregation, ScalarKwargs
     from narwhals._pandas_like.dataframe import PandasLikeDataFrame
     from narwhals._pandas_like.expr import PandasLikeExpr
+    from narwhals._utils import Implementation
+    from narwhals.dtypes import DType
+
+    NativeGroupBy: TypeAlias = "_NativeGroupBy[tuple[str, ...], Literal[True]]"
+    NativeSeriesGroupBy: TypeAlias = "_SeriesGroupBy[Any, tuple[str, ...]]"
+
+NativeApply: TypeAlias = "Callable[[pd.DataFrame], pd.Series[Any]]"
+InefficientNativeAggregation: TypeAlias = Literal["cov", "skew"]
+NativeAggregation: TypeAlias = Literal[
```

Comment:
I assume this is taken from pands docs. I wonder if it makes sense to keep only those that we can realistically hit from Narwhals, and add them as we increase the expressions supported